### PR TITLE
RFC: Rewrite homedir based on uv_os_homedir

### LINKED ISSUE #11331: don't rely on HOME environment variable for homedir?
In JuliaLang/IJulia.jl#309, there was a failure that seems to have been caused by the `HOME` environment variable not being defined (when Julia is launched by IPython in a virtualenv), although we are still trying to pin down the problem.

It occurred to me that relying on this environment variable for the `homedir()` function might not be completely robust.
- On Unix, we could instead call `getpwuid(getuid())->pw_dir`
- On Windows, there is [SHGetKnownFolderPath](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188%28v=vs.85%29.aspx) with the `FOLDERID_Profile` option.

We would probably have to write a helper routine in C to access these reliably (e.g. to get `struct passwd` from `pwd.h`).

### PR DESCRIPTION
**Pull request by ararslan**
Fixes #11331

Rather than relying on environment variables to determine the user's home directory, this calls `uv_os_homedir` from libuv, per @stevengj's suggestion.

The "magic number" of 1024 in there is what we're setting as the `PATH_MAX` if none is found in the system (see [src/support/dirpath.h](https://github.com/JuliaLang/julia/blob/master/src/support/dirpath.h)), so it seemed to me like a decent choice for the number of bytes to allocate for the output path. Let me know if that isn't reasonable or if there's a better choice.

For reference, here is the upstream libuv documentation on `uv_os_homedir`: http://docs.libuv.org/en/v1.x/misc.html#c.uv_os_homedir

### CODE REVIEW

File: base/path.jl [hunk 1]
```diff
@@ -57,7 +55,14 @@ splitdrive(path::AbstractString)
 
 Return the current user's home directory.
 """
-homedir()
+function homedir()
+    buf = Vector{Cchar}(1024)
+    sz = Ref{Csize_t}(sizeof(buf))
+    rc = ccall(:uv_os_homedir, Cint, (Ptr{Cchar}, Ptr{Csize_t}), buf, sz)
+    rc == 0 || error("unable to retrieve home directory")
```
Comment (yuyichao) [inline base/path.jl]: Hard to imagine it will happen but the correct way is to use `sz[] - 1` to resize the array.

File: base/path.jl [hunk 2]
```diff
@@ -57,7 +55,14 @@ splitdrive(path::AbstractString)
 
 Return the current user's home directory.
 """
-homedir()
+function homedir()
+    buf = Vector{Cchar}(1024)
+    sz = Ref{Csize_t}(sizeof(buf))
```
Comment (yuyichao) [inline base/path.jl]: Should be `sizeof(buf) + 1`

File: base/path.jl [hunk 3]
```diff
@@ -57,7 +55,14 @@ splitdrive(path::AbstractString)
...
+    rc = ccall(:uv_os_homedir, Cint, (Ptr{Cchar}, Ptr{Csize_t}), buf, sz)
+    rc == 0 || error("unable to retrieve home directory")
+    nbytes = Int(sz[])
+    return unsafe_string(pointer(buf[1:nbytes]))
```
Comment (yuyichao) [inline base/path.jl]: This is wrong and inefficient.

**Review by yuyichao — CHANGES_REQUESTED** [review]
And a better implementation is

```jl
julia> function homedir2()
           sz = Ref{Csize_t}(1024 + 1)
           buf = Vector{UInt8}(sz[] - 1)
           while true
               rc = ccall(:uv_os_homedir, Cint, (Ptr{UInt8}, Ptr{Csize_t}), buf, sz)
               if rc == 0
                   resize!(buf, sz[])
                   return String(buf)
               elseif rc == Base.UV_ENOBUFS # No Base. in base
                   resize!(buf, sz[] - 1)
               else
                   error("Unable to retrieve home directory.")
               end
           end
       end
```

**Comment by ararslan** [comment]
Thanks for the feedback, @yuyichao!

`String(buf)` causes an error because `Cchar` is `Int8`, not `UInt8`, so `String` can't convert the array. That's why I had initially used `unsafe_string` with `pointer`. Would it be better to convert to `UInt8`, then call the `String` constructor?

**Comment by yuyichao** [comment]
That s why I used UInt8

**Comment by ararslan** [comment]
Oh sorry! Missed that part.

**Comment by ararslan** [comment]
Looking better, @yuyichao?

**Review by yuyichao — APPROVED** [review]

**Comment by asampal** [comment]
With this change, is there any way to override the value that homedir() will return from the OS for homedir? I'd like my Julia homedir to be the home directory under my MSYS2/MinGW installation instead of the standard user dir Windows defines.

**Comment by stevengj** [comment]
@asampal, yes.  From the [libuv documentation](http://docs.libuv.org/en/v1.x/misc.html):

> Gets the current user’s home directory. On Windows, uv_os_homedir() first checks the USERPROFILE environment variable using GetEnvironmentVariableW(). If USERPROFILE is not set, GetUserProfileDirectoryW() is called. On all other operating systems, uv_os_homedir() first checks the HOME environment variable using getenv(3). If HOME is not set, getpwuid_r(3) is called. The user’s home directory is stored in buffer. When uv_os_homedir() is called, size indicates the maximum size of buffer. On success size is set to the string length of buffer. On UV_ENOBUFS failure size is set to the required length for buffer, including the null byte.

**Comment by asampal** [comment]
@stevengj if I redefine `USERPROFILE` in `juliarc.jl` I can get `homedir()` as I need it, but this will be overwritten by a new install.  I can't use `.juliarc.jl`, to set the home dir, obviously, and this isn't an env variable that should be changed globally (actually marked as read-only).  So without running Julia using a batch file, there's no other option?

**Pull request merged**